staging/src/k8s.io/apiserver/pkg/admission/cel/metrics.go

```diff
@@ -0,0 +1,105 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const (
+	metricsNamespace = "apiserver"
+	metricsSubsystem = "validating_admission_policy"
+)
+
+var (
+	// Metrics provides access to validation admission metrics.
+	Metrics = newValidationAdmissionMetrics()
+)
+
+// ValidatingAdmissionPolicyMetrics aggregates Prometheus metrics related to validation admission control.
+type ValidatingAdmissionPolicyMetrics struct {
+	policyCheck      *metrics.CounterVec
+	policyDefinition *metrics.CounterVec
+	policyLatency    *metrics.HistogramVec
+}
+
+func newValidationAdmissionMetrics() *ValidatingAdmissionPolicyMetrics {
+	check := metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      metricsNamespace,
+			Subsystem:      metricsSubsystem,
+			Name:           "check_total",
+			Help:           "Validation admission policy check total, labeled by policy and param resource, and further identified by binding, validation expression, enforcement action taken, and state.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"policy", "policy_binding", "validation_expression", "enforcement_action", "params", "state"},
+	)
+	definition := metrics.NewCounterVec(&metrics.CounterOpts{
+		Namespace:      metricsNamespace,
+		Subsystem:      metricsSubsystem,
+		Name:           "definition_count",
+		Help:           "Validation admission policy count total, labeled by state and enforcement action.",
+		StabilityLevel: metrics.ALPHA,
+	},
+		[]string{"state", "enforcement_action"},
+	)
+	latency := metrics.NewHistogramVec(&metrics.HistogramOpts{
+		Namespace: metricsNamespace,
+		Subsystem: metricsSubsystem,
+		Name:      "check_duration_seconds",
+		Help:      "Validation admission latency for individual validation expressions in seconds, labeled by policy and param resource, further including binding, state and enforcement action taken.",
+		// the bucket distribution here is based oo the benchmark suite at
+		// github.com/DangerOnTheRanger/cel-benchmark performed on 16-core Intel Xeon
+		// the lowest bucket was based around the 180ns/op figure for BenchmarkAccess,
+		// plus some additional leeway to account for the apiserver doing other things
+		// the largest bucket was chosen based on the fact that benchmarks indicate the
+		// same Xeon running a CEL expression close to the estimated cost limit takes
+		// around 760ms, so that bucket should only ever have the slowest CEL expressions
+		// in it
+		Buckets:        []float64{0.001, 0.01, 0.1, 1.0},
+		StabilityLevel: metrics.ALPHA,
+	},
+		[]string{"policy", "policy_binding", "validation_expression", "enforcement_action", "params", "state"},
```

Comment:
what's the cardinality on "params"?

Reply:
It should be equal to the number of [parameter resouces](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#api-shape) - my understanding is that number should be no higher than the cardinality of `policy_binding` (@jpbetz - pinging you here in case I'm off on this one).

Reply:
Correct, less than or equal to binding cardinality

Reply:
most of these labels seem like they would have concerningly high cardinality. "policy", "policy_binding", "params" are all names of instances of objects? so the cardinality is only limited to the number of instances of those types as someone creates in a cluster? what is validation_expression, the string expression, name of the validation rule that failed, index of the rule that failed, or something else?

Reply:
Right, those are all names (`validation_expression` is the name of the expression being checked). Do you think some of the labels should maybe be removed? I feel like the binding (and maybe the params?) label(s) could be removed without too much impact towards debugging.

Reply:
My understanding was that metric labels tied to object names were undesirable since they have ~unbounded cardinality.

_If_ we were going to have a name as a label, the binding name identifies the policy/params tuple (at least until singleton policies that are effective without a binding get implemented)

Reply:
**If** the label cardinality is truly capped to 100s per cluster, then even if the total cardinality is unbounded, it should not pose a terrible problem. We have far worse issues with the resource label for the apiserver request metrics. It basically just means we're not going to be able to do meaningful aggregation across multiple clusters over these dimensions. If the cardinality for a single cluster is actually unbounded, then I would highly recommend against these labels.

Reply:
I'd consider this equivalent to the `resource` label in API server request metrics... every new CRD spawns a new `resource` value, and every new policy or policy_binding instance (depending on which of these labels was selected) would spawn a new label value

Reply:
It's probably fine then, moderately discouraged but if it would help debugging then I'd be okay with it.